tools/update.php

```diff
@@ -402,6 +460,92 @@ function applyPatches($patches, $db_config, $apply_patches = false) : bool
     return true;
 }
 
+function getPrivileges($db_config): array {
+    $A = $db_config['database'];
+    $u = $db_config['username'];
+    $h = $db_config['host'];
+    $p = $db_config['password'];
+    /* Run 'SHOW privileges' for this user and parse the output. We are passing
+     * the password value on bhelaf of the user here as the SHOW command does
+     * not alter any data.
+     *
+     * NOTE this may have issues when dealing with passwords that contain
+     * special characters that the shell may interpret differently. TBD.
+     */
+    $cmd = "echo 'show privileges' "
+        . "| mysql -h $h -u $u --password='$p' -A $A "
+        . "| cut -f 1";
+    return array_map('strtolower',
+        explode(
+            "\n", 
+            /* Don't use doExec() here because we are entering the password and 
+             * don't want it printed to stdout.
+             */
+            shell_exec($cmd)
+        )
+    );
+}
+
+
+/**
+ * Using the keywords from a MySQL script, determine whether the user in
+ * $db_config is able to execute all the commands corresponding to the keywords.
+ */
+function canApplyPatch(string $patch, $db_config): bool
+{
+    $num_missing = count(
+        getMissingPrivileges(
+            $patch,
+            $db_config
+        )
+    );
+    return $num_missing === 0;
+}
+
+/**
+ * Use bash commands to extract MySQL keywords from an sql script.
+ */
+function getKeywords(string $patch): array
+{
+    //Scan file for MySQL keywords. Exclude comments and parentheses using grep.
+    $cmd = "cut $patch -d ' ' -f 1 "
+        . "| sort -u "
+        . "| grep -v '\--' | grep -v ')'";
+    /* Processing chain: 
+     *     Split output of shell exec by new lines into an array
+     *     Convert all array values to lower case
+     *     Filter out any false values, empty strings in this case.
+     */
+    return array_filter(
+        array_map(
+            'strtolower',
+            explode(
+                "\n", 
+                /* Don't use doExec() here because we are entering the password
+                 * and don't want it printed to stdout.
+                 */
+                shell_exec($cmd)
+            )
+        )
+    );
+}
+function getMissingPrivileges(string $patch, $db_config) {
+    $keywords = getKeywords($patch);
+    $privileges = getPrivileges($db_config);
+    $irrelevantWords = array(
```

Comment:
should also check for upper case as in_array is case sensitive

Reply:
A few lines above this I used `array_map` to convert all `$keywords` to lower case which should take care of this.